File: generate_data.py
import pandas as pd
import numpy as np
import os
import random

# Set seed for reproducibility
np.random.seed(42)
random.seed(42)

def generate_clinical_data(n=200):
    print("Generating Clinical Data...")
    ids = [f"P{i:03d}" for i in range(1, n+1)]
    age = np.random.normal(55, 12, n).astype(int)
    gender = np.random.choice(['Male', 'Female'], n)
    # Simulate BMI with some missing values
    bmi = np.random.normal(25, 4, n)
    bmi[np.random.choice(n, int(n*0.05), replace=False)] = np.nan # 5% missing
    
    # Blood pressure correlated with age and BMI
    sys_bp = 100 + 0.5 * age + 0.8 * np.nan_to_num(bmi, nan=25) + np.random.normal(0, 10, n)
    
    # Diagnosis (Target): Higher risk if older or high BMI
    interaction = (age - 50)/10 + (np.nan_to_num(bmi, nan=25) - 25)/5
    prob = 1 / (1 + np.exp(-(interaction - 1)))
    diagnosis = [1 if p > 0.5 else 0 for p in prob]
    
    df = pd.DataFrame({
        'PatientID': ids,
        'Age': age,
        'Gender': gender,
        'BMI': np.round(bmi, 1),
        'SystolicBP': np.round(sys_bp, 0),
        'Diagnosis': diagnosis
    })
    
    # Introduce some "dirty" data for cleaning exercises
    # 1. Invalid age
    df.loc[0, 'Age'] = 150
    # 2. Typos in Gender
    df.loc[1, 'Gender'] = 'Femle'
    
    df.to_csv('data/clinical/simulated_ehr.csv', index=False)
    print("Saved data/clinical/simulated_ehr.csv")

def generate_tcm_data(n_transactions=500):
    print("Generating TCM Data...")
    # Common TCM herbs
    herbs = [
        'GanCao', 'RenShen', 'HuangQi', 'BaiZhu', 'FuLing', 
        'DangGui', 'ChuanXiong', 'BaiShao', 'ShuDiHuang', 
        'ChaiHu', 'GuiZhi', 'MaHuang', 'XingRen', 'ShiGao'
    ]
    
    # Create some strong associations (rules)
    # Rule 1: GuiZhi + MaHuang (often appear together)
    # Rule 2: SiJunZiTang (RenShen, BaiZhu, FuLing, GanCao)
    
    transactions = []
    for i in range(n_transactions):
        # Base random selection
        n_items = random.randint(3, 8)
        selected = random.sample(herbs, n_items)
        selected_set = set(selected)
        
        # Inject patterns
        r = random.random()
        if r < 0.3: # 30% chance for MaHuang-GuiZhi pair
            selected_set.add('MaHuang')
            selected_set.add('GuiZhi')
        elif r < 0.6: # 30% chance for SiJunZiTang subset partial
            selected_set.update(['RenShen', 'BaiZhu', 'FuLing'])
            
        transactions.append(",".join(list(selected_set)))
        
    df = pd.DataFrame({'PrescriptionID': range(1, n_transactions+1), 'Herbs': transactions})
    df.to_csv('data/tcm/tcm_prescriptions.csv', index=False)
    print("Saved data/tcm/tcm_prescriptions.csv")

def generate_omics_data(n_genes=1000, n_samples=20):
    print("Generating Omics Data...")
    # Simple DE analysis simulation
    # 10 Control, 10 Disease
    samples = [f"Ctrl_{i}" for i in range(1, 11)] + [f"Dis_{i}" for i in range(1, 11)]
    groups = ['Control'] * 10 + ['Disease'] * 10
    
    # Base expression log-normal
    expr = np.random.normal(10, 2, (n_genes, n_samples))
    
    # Make first 50 genes differentially expressed (Up-regulated in Disease)
    expr[:50, 10:] += 2 
    # Make next 50 genes down-regulated in Disease
    expr[50:100, 10:] -= 2
    
    genes = [f"Gene_{i:04d}" for i in range(1, n_genes+1)]
    
    df_expr = pd.DataFrame(expr, columns=samples, index=genes)
    df_expr.index.name = 'GeneID'
    
    df_meta = pd.DataFrame({'SampleID': samples, 'Group': groups})
    
    df_expr.to_csv('data/omics/gene_expression.csv')
    df_meta.to_csv('data/omics/sample_info.csv', index=False)
    print("Saved data/omics/gene_expression.csv and sample_info.csv")

if __name__ == "__main__":
    generate_clinical_data()
    generate_tcm_data()
    generate_omics_data()

File: R/generate_data.R
#' Generate Simulated Medical Data
#'
#' This script generates simulated datasets for the course:
#' 1. Clinical Data (EHR simulation)
#' 2. TCM Prescription Data (Association rules)
#' 3. Omics Data (Differential expression)

set.seed(42)

# --- 1. Clinical Data ---
generate_clinical <- function(n = 200) {
  message("Generating Clinical Data...")
  
  ids <- sprintf("P%03d", 1:n)
  age <- as.integer(rnorm(n, mean = 55, sd = 12))
  gender <- sample(c("Male", "Female"), n, replace = TRUE)
  
  # BMI with missing values
  bmi <- rnorm(n, mean = 25, sd = 4)
  # 5% missing
  bmi[sample(1:n, size = n * 0.05)] <- NA
  
  # Blood pressure
  bmi_fill <- ifelse(is.na(bmi), 25, bmi)
  sys_bp <- 100 + 0.5 * age + 0.8 * bmi_fill + rnorm(n, 0, 10)
  
  # Diagnosis Prob
  interaction <- (age - 50)/10 + (bmi_fill - 25)/5
  prob <- 1 / (1 + exp(-(interaction - 1)))
  diagnosis <- ifelse(prob > 0.5, 1, 0)
  
  df <- data.frame(
    PatientID = ids,
    Age = age,
    Gender = gender,
    BMI = round(bmi, 1),
    SystolicBP = round(sys_bp, 0),
    Diagnosis = diagnosis
  )
  
  # Add "dirty" data
  df$Age[1] <- 150
  df$Gender[2] <- "Femle"
  
  if(!dir.exists("data/clinical")) dir.create("data/clinical", recursive = TRUE)
  write.csv(df, "data/clinical/simulated_ehr.csv", row.names = FALSE)
  message("Saved data/clinical/simulated_ehr.csv")
}

# --- 2. TCM Data ---
generate_tcm <- function(n_transactions = 500) {
  message("Generating TCM Data...")
  herbs <- c('GanCao', 'RenShen', 'HuangQi', 'BaiZhu', 'FuLing', 
             'DangGui', 'ChuanXiong', 'BaiShao', 'ShuDiHuang', 
             'ChaiHu', 'GuiZhi', 'MaHuang', 'XingRen', 'ShiGao')
  
  transactions <- character(n_transactions)
  
  for(i in 1:n_transactions) {
    n_items <- sample(3:8, 1)
    selected <- sample(herbs, n_items)
    unique_selected <- unique(selected)
    
    # Inject patterns
    r <- runif(1)
    if(r < 0.3) {
      unique_selected <- unique(c(unique_selected, 'MaHuang', 'GuiZhi'))
    } else if (r < 0.6) {
      unique_selected <- unique(c(unique_selected, 'RenShen', 'BaiZhu', 'FuLing'))
    }
    
    transactions[i] <- paste(unique_selected, collapse = ",")
  }
  
  df <- data.frame(PrescriptionID = 1:n_transactions, Herbs = transactions)
  
  if(!dir.exists("data/tcm")) dir.create("data/tcm", recursive = TRUE)
  write.csv(df, "data/tcm/tcm_prescriptions.csv", row.names = FALSE)
  message("Saved data/tcm/tcm_prescriptions.csv")
}

# --- 3. Omics Data ---
generate_omics <- function(n_genes = 1000, n_samples = 20) {
  message("Generating Omics Data...")
  samples <- c(paste0("Ctrl_", 1:10), paste0("Dis_", 1:10))
  groups <- c(rep("Control", 10), rep("Disease", 10))
  
  expr <- matrix(rnorm(n_genes * n_samples, mean = 10, sd = 2), 
                 nrow = n_genes, ncol = n_samples)
  rownames(expr) <- sprintf("Gene_%04d", 1:n_genes)
  colnames(expr) <- samples
  
  # Differential Expression
  # First 50 up in Disease
  expr[1:50, 11:20] <- expr[1:50, 11:20] + 2
  # Next 50 down in Disease
  expr[51:100, 11:20] <- expr[51:100, 11:20] - 2
  
  if(!dir.exists("data/omics")) dir.create("data/omics", recursive = TRUE)
  write.csv(expr, "data/omics/gene_expression.csv") # keep rownames
  
  meta <- data.frame(SampleID = samples, Group = groups)
  write.csv(meta, "data/omics/sample_info.csv", row.names = FALSE)
  message("Saved data/omics/gene_expression.csv and sample_info.csv")
}

# Run generation
generate_clinical()
generate_tcm()
generate_omics()
